File: ios/RezumateNative/Services/ResumeParser.swift
import Foundation

struct ResumeParser {

    private struct DetectedSection {
        let key: String
        let title: String
    }

    private static let sectionAliases: [String: String] = [
        "summary": "summary",
        "objective": "summary",
        "professional summary": "summary",
        "career summary": "summary",
        "profile": "summary",
        "about": "summary",
        "experience": "experience",
        "work experience": "experience",
        "professional experience": "experience",
        "employment": "experience",
        "employment history": "experience",
        "work history": "experience",
        "projects": "projects",
        "personal projects": "projects",
        "project experience": "projects",
        "side projects": "projects",
        "open source": "projects",
        "education": "education",
        "academic background": "education",
        "academic history": "education",
        "skills": "skills",
        "technical skills": "skills",
        "technologies": "skills",
        "core competencies": "skills",
        "technical competencies": "skills",
        "certifications": "certifications",
        "awards": "awards",
        "publications": "publications",
        "volunteer": "volunteer",
        "volunteering": "volunteer",
    ]

    // Known single-word or short location indicators
    private static let knownLocations: Set<String> = [
        "remote", "india", "usa", "us", "uk", "united kingdom", "united states",
        "new york", "san francisco", "los angeles", "london", "berlin", "toronto",
        "singapore", "australia", "canada", "germany", "france", "europe", "asia",
        "worldwide", "globally", "anywhere", "hybrid"
    ]

    // Compiled date-range regex (matches "Jan. 2022 -- Present", "2019 - 2022", etc.)
    private static let dateRegex: NSRegularExpression? = {
        let pattern =
            #"(?:(?:Jan(?:uary)?|Feb(?:ruary)?|Mar(?:ch)?|Apr(?:il)?|May|Jun(?:e)?|Jul(?:y)?|Aug(?:ust)?|Sep(?:t(?:ember)?)?|Oct(?:ober)?|Nov(?:ember)?|Dec(?:ember)?)\.?\s+)?"# +
            #"\d{4}\s*[-–—]+\s*(?:Present|Current|\d{4}|(?:Jan(?:uary)?|Feb(?:ruary)?|Mar(?:ch)?|Apr(?:il)?|May|Jun(?:e)?|Jul(?:y)?|Aug(?:ust)?|Sep(?:t(?:ember)?)?|Oct(?:ober)?|Nov(?:ember)?|Dec(?:ember)?)\.?\s+\d{4})"#
        return try? NSRegularExpression(pattern: pattern, options: .caseInsensitive)
    }()

    // MARK: - Public

    static func parse(_ text: String) -> ResumeDocument {
        var doc = ResumeDocument()
        let lines = text
            .replacingOccurrences(of: "\r\n", with: "\n")
            .replacingOccurrences(of: "\r", with: "\n")
            .components(separatedBy: "\n")
            .map { $0.trimmingCharacters(in: .whitespaces) }

        guard !lines.isEmpty else { return doc }

        var idx = 0

        // Skip leading blanks, then grab name
        while idx < lines.count, lines[idx].isEmpty { idx += 1 }
        guard idx < lines.count else { return doc }
        doc.name = lines[idx]
        idx += 1

        // Contact block: lines before the first section header
        while idx < lines.count {
            let line = lines[idx]
            if line.isEmpty { idx += 1; continue }
            if detectedSection(for: line) != nil { break }
            if !parseContactLine(line, into: &doc) {
                doc.unmappedContent.append(line)
            }
            idx += 1
        }

        // Section parsing via FSM
        var currentSection: DetectedSection? = nil
        var sectionContent: [String] = []

        while idx < lines.count {
            let line = lines[idx]
            idx += 1

            if let section = detectedSection(for: line) {
                if let prev = currentSection {
                    processSection(prev, lines: sectionContent, into: &doc)
                }
                currentSection = section
                sectionContent = []
            } else {
                sectionContent.append(line)
            }
        }

        if let section = currentSection {
            processSection(section, lines: sectionContent, into: &doc)
        }

        return doc
    }

    // MARK: - Section detection

    private static func detectedSection(for line: String) -> DetectedSection? {
        let trimmed = line.trimmingCharacters(in: .whitespaces)
        guard !trimmed.isEmpty, trimmed.count >= 3, trimmed.count <= 50 else { return nil }

        let lower = trimmed.lowercased()
        if let canonical = sectionAliases[lower] {
            return DetectedSection(key: canonical, title: trimmed)
        }

        // All-uppercase line (no digits, only letters and spaces)
        let scalars = trimmed.unicodeScalars
        let letters = scalars.filter { CharacterSet.letters.contains($0) }
        guard !letters.isEmpty else { return nil }
        let allUpper = letters.allSatisfy { CharacterSet.uppercaseLetters.contains($0) }
        if allUpper {
            return DetectedSection(key: sectionAliases[lower] ?? lower, title: trimmed)
        }

        return nil
    }

    // MARK: - Contact parsing

    @discardableResult
    private static func parseContactLine(_ line: String, into doc: inout ResumeDocument) -> Bool {
        let parts = line
            .components(separatedBy: CharacterSet(charactersIn: "|•·$"))
            .map { $0.trimmingCharacters(in: .whitespaces) }
            .filter { !$0.isEmpty }

        var parsedAny = false
        for part in (parts.count > 1 ? parts : [line]) {
            let lower = part.lowercased()

            if doc.email == nil, let email = extractEmail(from: part) {
                doc.email = email
                parsedAny = true
            } else if lower.contains("linkedin.com/in/") {
                doc.linkedin = doc.linkedin ?? (extractLinkedIn(from: part) ?? part)
                parsedAny = true
            } else if lower.contains("github.com/") {
                doc.github = doc.github ?? (extractGitHub(from: part) ?? part)
                parsedAny = true
            } else if lower.hasPrefix("http") || lower.hasPrefix("www.")
                || lower.contains(".dev") || lower.contains(".io")
                || (lower.contains(".com") && !lower.contains("@"))
                || (lower.contains(".app") && !lower.contains("@")) {
                if doc.website == nil { doc.website = part }
                parsedAny = true
            } else if doc.phone == nil, let phone = extractPhone(from: part) {
                doc.phone = phone
                parsedAny = true
            } else if isLikelyLocation(part) {
                if doc.location == nil { doc.location = part }
                parsedAny = true
            }
        }
        return parsedAny
    }

    private static func extractEmail(from text: String) -> String? {
        let pattern = #"[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}"#
        guard let rx = try? NSRegularExpression(pattern: pattern),
              let m = rx.firstMatch(in: text, range: NSRange(text.startIndex..., in: text)),
              let r = Range(m.range, in: text) else { return nil }
        return String(text[r])
    }

    private static func extractPhone(from text: String) -> String? {
        let pattern = #"(?<![A-Za-z0-9])\+?[0-9(][0-9\s().-]{8,}[0-9](?![A-Za-z0-9])"#
        guard let rx = try? NSRegularExpression(pattern: pattern) else { return nil }

        for match in rx.matches(in: text, range: NSRange(text.startIndex..., in: text)) {
            guard let range = Range(match.range, in: text) else { continue }
            let candidate = String(text[range]).trimmingCharacters(in: .whitespacesAndNewlines)
            let digitCount = candidate.unicodeScalars.filter { CharacterSet.decimalDigits.contains($0) }.count
            if (10...15).contains(digitCount) {
                return candidate
            }
        }
        return nil
    }

    private static func extractLinkedIn(from text: String) -> String? {
        let pattern = #"linkedin\.com/in/([a-zA-Z0-9\-]+)"#
        guard let rx = try? NSRegularExpression(pattern: pattern, options: .caseInsensitive),
              let m = rx.firstMatch(in: text, range: NSRange(text.startIndex..., in: text)),
              m.numberOfRanges > 1,
              let r = Range(m.range(at: 1), in: text) else { return nil }
        return String(text[r])
    }

    private static func extractGitHub(from text: String) -> String? {
        let pattern = #"github\.com/([a-zA-Z0-9\-]+)"#
        guard let rx = try? NSRegularExpression(pattern: pattern, options: .caseInsensitive),
              let m = rx.firstMatch(in: text, range: NSRange(text.startIndex..., in: text)),
              m.numberOfRanges > 1,
              let r = Range(m.range(at: 1), in: text) else { return nil }
        return String(text[r])
    }

    private static func isLikelyLocation(_ text: String) -> Bool {
        let lower = text.lowercased().trimmingCharacters(in: .whitespaces)
        if knownLocations.contains(lower) { return true }
        // "City, ST" pattern
        if lower.range(of: #"^[a-z\s]+,\s*[a-z]{2}$"#, options: .regularExpression) != nil { return true }
        // Short, title-cased, no special chars means likely a city/country.
        let words = text.components(separatedBy: .whitespaces).filter { !$0.isEmpty }
        if words.count <= 3 && !text.contains("@") && !text.contains(".") && !text.contains("/") {
            let allCapped = words.allSatisfy { $0.first?.isUppercase == true }
            if allCapped && words.count == 1 && text.count >= 3 { return true }
        }
        return false
    }

    // MARK: - Section processing

    private static func processSection(_ section: DetectedSection, lines: [String], into doc: inout ResumeDocument) {
        switch section.key {
        case "summary":
            doc.summary = lines.filter { !$0.isEmpty }.joined(separator: " ")
        case "experience":
            doc.experience = parseExperienceEntries(lines)
        case "projects":
            doc.projects = parseProjectEntries(lines)
        case "education":
            doc.education = parseEducationEntries(lines)
        case "skills":
            doc.skillCategories = parseSkillsSection(lines)
        default:
            let meaningfulLines = lines.filter { !$0.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }
            if !meaningfulLines.isEmpty {
                doc.additionalSections.append(
                    AdditionalResumeSection(title: section.title, lines: meaningfulLines)
                )
            }
        }
    }

    // MARK: - Experience

    private static func parseExperienceEntries(_ lines: [String]) -> [ExperienceEntry] {
        var entries: [ExperienceEntry] = []
        var current: ExperienceEntry? = nil
        var expectTitle = false

        for line in lines {
            if line.isEmpty { continue }

            if let bullet = asBullet(line) {
                current?.bullets.append(bullet)
                expectTitle = false
                continue
            }

            if current == nil {
                var e = ExperienceEntry()
                let (name, date) = splitNameAndDate(line)
                e.company = name
                e.dateRange = date
                current = e
                expectTitle = true
            } else if expectTitle {
                let (title, loc) = splitTitleAndLocation(line)
                current!.title = title
                current!.location = loc
                expectTitle = false
            } else if shouldContinuePreviousBullet(line, current: current) {
                var updated = current!
                appendToLastBullet(line, entry: &updated)
                current = updated
            } else {
                entries.append(current!)
                var e = ExperienceEntry()
                let (name, date) = splitNameAndDate(line)
                e.company = name
                e.dateRange = date
                current = e
                expectTitle = true
            }
        }

        if let last = current { entries.append(last) }
        return entries
    }

    // MARK: - Projects

    private static func parseProjectEntries(_ lines: [String]) -> [ProjectEntry] {
        var entries: [ProjectEntry] = []
        var current: ProjectEntry? = nil

        for line in lines {
            if line.isEmpty { continue }

            if let bullet = asBullet(line) {
                if current == nil { current = ProjectEntry(name: "") }
                current!.bullets.append(bullet)
                continue
            }

            if shouldContinuePreviousProjectBullet(line, current: current) {
                var updated = current!
                appendToLastProjectBullet(line, entry: &updated)
                current = updated
                continue
            }

            if let prev = current { entries.append(prev) }
            // Strip trailing date range from project name line
            let (name, _) = splitNameAndDate(line)
            current = ProjectEntry(name: name.isEmpty ? line : name)
        }

        if let last = current { entries.append(last) }
        return entries
    }

    // MARK: - Education

    private static func parseEducationEntries(_ lines: [String]) -> [EducationEntry] {
        var entries: [EducationEntry] = []
        var current: EducationEntry? = nil
        var expectDegree = false

        for line in lines {
            if line.isEmpty { continue }

            if let bullet = asBullet(line) {
                current?.details.append(bullet)
                expectDegree = false
                continue
            }

            if current == nil {
                var e = EducationEntry()
                let (inst, date) = splitNameAndDate(line)
                e.institution = inst
                e.dateRange = date
                current = e
                expectDegree = true
            } else if expectDegree {
                let (deg, loc) = splitTitleAndLocation(line)
                current!.degree = deg
                current!.location = loc
                expectDegree = false
            } else {
                entries.append(current!)
                var e = EducationEntry()
                let (inst, date) = splitNameAndDate(line)
                e.institution = inst
                e.dateRange = date
                current = e
                expectDegree = true
            }
        }

        if let last = current { entries.append(last) }
        return entries
    }

    // MARK: - Skills

    private static func parseSkillsSection(_ lines: [String]) -> [SkillCategory] {
        var cats: [SkillCategory] = []

        for line in lines {
            if line.isEmpty { continue }
            if let colonIdx = line.firstIndex(of: ":") {
                let name = String(line[..<colonIdx]).trimmingCharacters(in: .whitespaces)
                let items = String(line[line.index(after: colonIdx)...]).trimmingCharacters(in: .whitespaces)
                if !name.isEmpty && !items.isEmpty && name.count < 40 {
                    cats.append(SkillCategory(name: name, items: items))
                    continue
                }
            }
            cats.append(SkillCategory(name: "", items: line))
        }

        return cats
    }

    // MARK: - Helpers

    private static func asBullet(_ line: String) -> String? {
        let t = line.trimmingCharacters(in: .whitespaces)
        let bulletChars: Set<Character> = ["•", "·", "▪", "▸", "◦", "‣", "–", "—"]
        if let first = t.first, bulletChars.contains(first) {
            return String(t.dropFirst()).trimmingCharacters(in: .whitespaces)
        }
        if t.hasPrefix("- ") { return String(t.dropFirst(2)).trimmingCharacters(in: .whitespaces) }
        if t.hasPrefix("* ") { return String(t.dropFirst(2)).trimmingCharacters(in: .whitespaces) }
        return nil
    }

    private static func shouldContinuePreviousBullet(_ line: String, current: ExperienceEntry?) -> Bool {
        guard let current, !current.bullets.isEmpty else { return false }
        let (_, date) = splitNameAndDate(line)
        return date.isEmpty
    }

    private static func appendToLastBullet(_ line: String, entry: inout ExperienceEntry) {
        guard !entry.bullets.isEmpty else { return }
        let continuation = line.trimmingCharacters(in: .whitespaces)
        guard !continuation.isEmpty else { return }
        entry.bullets[entry.bullets.count - 1] += " " + continuation
    }

    private static func shouldContinuePreviousProjectBullet(_ line: String, current: ProjectEntry?) -> Bool {
        guard let current, !current.bullets.isEmpty else { return false }
        let trimmed = line.trimmingCharacters(in: .whitespaces)
        guard !trimmed.isEmpty else { return false }
        if splitNameAndDate(trimmed).1.isEmpty == false { return false }
        if trimmed.count > 38 { return true }
        if let first = trimmed.first, first.isLowercase || "(),;:/".contains(first) { return true }
        return false
    }

    private static func appendToLastProjectBullet(_ line: String, entry: inout ProjectEntry) {
        guard !entry.bullets.isEmpty else { return }
        let continuation = line.trimmingCharacters(in: .whitespaces)
        guard !continuation.isEmpty else { return }
        entry.bullets[entry.bullets.count - 1] += " " + continuation
    }

    // Splits "Company Name Feb. 2026 -- Present" -> ("Company Name", "Feb. 2026 -- Present")
    private static func splitNameAndDate(_ line: String) -> (String, String) {
        guard let regex = dateRegex else { return (line, "") }

        var lastMatch: NSTextCheckingResult? = nil
        regex.enumerateMatches(in: line, range: NSRange(line.startIndex..., in: line)) { m, _, _ in
            lastMatch = m
        }

        guard let match = lastMatch, let matchRange = Range(match.range, in: line) else {
            return (line, "")
        }

        let date = String(line[matchRange]).trimmingCharacters(in: .whitespaces)
        let name = String(line[..<matchRange.lowerBound]).trimmingCharacters(in: .whitespaces)
        return name.isEmpty ? (line, "") : (name, date)
    }

    // Splits "Founder & Lead Engineer Remote" -> ("Founder & Lead Engineer", "Remote")
    private static func splitTitleAndLocation(_ line: String) -> (String, String) {
        let words = line.components(separatedBy: .whitespaces).filter { !$0.isEmpty }
        guard words.count > 1 else { return (line, "") }

        // Preserve a complete "City, ST" suffix. PDF text extraction commonly removes
        // column spacing, so a row such as "iOS Engineer  San Francisco, CA" arrives as
        // one line. Prefer a known multi-word city, then fall back to the last word.
        if let comma = line.lastIndex(of: ",") {
            let state = line[line.index(after: comma)...].trimmingCharacters(in: .whitespaces)
            let beforeComma = line[..<comma].trimmingCharacters(in: .whitespaces)
            if state.range(of: #"^[A-Za-z]{2}$"#, options: .regularExpression) != nil {
                let lowerBeforeComma = beforeComma.lowercased()
                let knownCity = knownLocations
                    .sorted { $0.count > $1.count }
                    .first { lowerBeforeComma == $0 || lowerBeforeComma.hasSuffix(" \($0)") }

                let city: String
                let rawTitle: String
                if let knownCity {
                    city = String(beforeComma.suffix(knownCity.count))
                    rawTitle = String(beforeComma.dropLast(knownCity.count))
                } else if let cityStart = beforeComma.lastIndex(where: { $0.isWhitespace }) {
                    city = String(beforeComma[beforeComma.index(after: cityStart)...])
                    rawTitle = String(beforeComma[..<cityStart])
                } else {
                    city = ""
                    rawTitle = ""
                }

                let title = rawTitle.trimmingCharacters(in: CharacterSet(charactersIn: " |–—-\t"))
                if !title.isEmpty, !city.isEmpty {
                    return (title, "\(city), \(state.uppercased())")
                }
            }
        }

        if let last = words.last, knownLocations.contains(last.lowercased()) {
            let title = words.dropLast().joined(separator: " ")
            return (title.isEmpty ? line : title, last)
        }
        if words.count >= 3 {
            let lastTwo = words.suffix(2).joined(separator: " ")
            if knownLocations.contains(lastTwo.lowercased()) {
                let title = words.dropLast(2).joined(separator: " ")
                return (title.isEmpty ? line : title, lastTwo)
            }
        }
        return (line, "")
    }
}

File: android/app/src/main/java/com/aftaab/rezumate/domain/ResumeParser.kt
package com.aftaab.rezumate.domain

import com.aftaab.rezumate.model.AdditionalResumeSection
import com.aftaab.rezumate.model.EducationEntry
import com.aftaab.rezumate.model.ExperienceEntry
import com.aftaab.rezumate.model.ProjectEntry
import com.aftaab.rezumate.model.ResumeDocument
import com.aftaab.rezumate.model.SkillCategory

object ResumeParser {
    private val sectionAliases = mapOf(
        "summary" to "summary",
        "objective" to "summary",
        "professional summary" to "summary",
        "career summary" to "summary",
        "profile" to "summary",
        "about" to "summary",
        "experience" to "experience",
        "work experience" to "experience",
        "professional experience" to "experience",
        "employment" to "experience",
        "employment history" to "experience",
        "work history" to "experience",
        "projects" to "projects",
        "personal projects" to "projects",
        "project experience" to "projects",
        "side projects" to "projects",
        "open source" to "projects",
        "education" to "education",
        "academic background" to "education",
        "academic history" to "education",
        "skills" to "skills",
        "technical skills" to "skills",
        "technologies" to "skills",
        "core competencies" to "skills",
        "technical competencies" to "skills",
        "certifications" to "certifications",
        "awards" to "awards",
        "publications" to "publications",
        "volunteer" to "volunteer",
        "volunteering" to "volunteer",
    )

    private val knownLocations = setOf(
        "remote", "india", "usa", "us", "uk", "united kingdom", "united states",
        "new york", "san francisco", "los angeles", "london", "berlin", "toronto",
        "singapore", "australia", "canada", "germany", "france", "europe", "asia",
        "worldwide", "globally", "anywhere", "hybrid",
    )

    private val dateRegex = Regex(
        "(?:(?:Jan(?:uary)?|Feb(?:ruary)?|Mar(?:ch)?|Apr(?:il)?|May|Jun(?:e)?|Jul(?:y)?|Aug(?:ust)?|Sep(?:t(?:ember)?)?|Oct(?:ober)?|Nov(?:ember)?|Dec(?:ember)?)\\.?\\s+)?" +
            "\\d{4}\\s*[-–—]+\\s*(?:Present|Current|\\d{4}|(?:Jan(?:uary)?|Feb(?:ruary)?|Mar(?:ch)?|Apr(?:il)?|May|Jun(?:e)?|Jul(?:y)?|Aug(?:ust)?|Sep(?:t(?:ember)?)?|Oct(?:ober)?|Nov(?:ember)?|Dec(?:ember)?)\\.?\\s+\\d{4})",
        RegexOption.IGNORE_CASE,
    )

    fun parse(text: String): ResumeDocument {
        val document = ResumeDocument()
        val lines = splitPreservingEmpty(text.replace("\r\n", "\n").replace("\r", "\n"))
            .map { it.trimHorizontalWhitespace() }
        var index = 0
        while (index < lines.size && lines[index].isEmpty()) index++
        if (index >= lines.size) return document

        document.name = lines[index++]
        while (index < lines.size) {
            val line = lines[index]
            if (line.isEmpty()) {
                index++
                continue
            }
            if (sectionName(line) != null) break
            if (!parseContactLine(line, document)) {
                document.unmappedContent += line
            }
            index++
        }

        var currentSection: String? = null
        var sectionContent = mutableListOf<String>()
        while (index < lines.size) {
            val line = lines[index++]
            val section = sectionName(line)
            if (section != null) {
                currentSection?.let { processSection(it, sectionContent, document) }
                currentSection = section
                sectionContent = mutableListOf()
            } else {
                sectionContent += line
            }
        }
        currentSection?.let { processSection(it, sectionContent, document) }
        return document
    }

    private fun sectionName(line: String): String? {
        val trimmed = line.trimHorizontalWhitespace()
        if (trimmed.isEmpty() || trimmed.length !in 3..50) return null
        val lower = trimmed.lowercase()
        sectionAliases[lower]?.let { return it }
        val letters = trimmed.filter(Char::isLetter)
        if (letters.isEmpty()) return null
        return if (letters.all(Char::isUpperCase)) sectionAliases[lower] ?: lower else null
    }

    private fun parseContactLine(line: String, document: ResumeDocument): Boolean {
        val splitParts = line.split(Regex("[|•·${'$'}]"))
            .map { it.trimHorizontalWhitespace() }
            .filter(String::isNotEmpty)
        var parsedAny = false
        for (part in if (splitParts.size > 1) splitParts else listOf(line)) {
            val lower = part.lowercase()
            val email = extractEmail(part)
            val phone = extractPhone(part)
            when {
                document.email == null && email != null -> {
                    document.email = email
                    parsedAny = true
                }
                "linkedin.com/in/" in lower -> {
                    if (document.linkedin == null) document.linkedin = extractLinkedIn(part) ?: part
                    parsedAny = true
                }
                "github.com/" in lower -> {
                    if (document.github == null) document.github = extractGitHub(part) ?: part
                    parsedAny = true
                }
                lower.startsWith("http") || lower.startsWith("www.") ||
                    ".dev" in lower || ".io" in lower ||
                    (".com" in lower && "@" !in lower) ||
                    (".app" in lower && "@" !in lower) -> {
                    if (document.website == null) document.website = part
                    parsedAny = true
                }
                document.phone == null && phone != null -> {
                    document.phone = phone
                    parsedAny = true
                }
                isLikelyLocation(part) -> {
                    if (document.location == null) document.location = part
                    parsedAny = true
                }
            }
        }
        return parsedAny
    }

    private fun extractEmail(text: String): String? =
        Regex("[a-zA-Z0-9._%+\\-]+@[a-zA-Z0-9.\\-]+\\.[a-zA-Z]{2,}").find(text)?.value

    private fun extractPhone(text: String): String? {
        val matches = Regex("""(?<![A-Za-z0-9])\+?[0-9(][0-9\s().-]{8,}[0-9](?![A-Za-z0-9])""").findAll(text)
        for (match in matches) {
            val candidate = match.value.trim()
            val digits = candidate.count { it.isDigit() }
            if (digits in 10..15) return candidate
        }
        return null
    }

    private fun extractLinkedIn(text: String): String? =
        Regex("linkedin\\.com/in/([a-zA-Z0-9\\-]+)", RegexOption.IGNORE_CASE)
            .find(text)?.groupValues?.get(1)

    private fun extractGitHub(text: String): String? =
        Regex("github\\.com/([a-zA-Z0-9\\-]+)", RegexOption.IGNORE_CASE)
            .find(text)?.groupValues?.get(1)

    private fun isLikelyLocation(text: String): Boolean {
        val lower = text.lowercase().trimHorizontalWhitespace()
        if (lower in knownLocations) return true
        if (Regex("^[a-z\\s]+,\\s*[a-z]{2}${'$'}").containsMatchIn(lower)) return true
        val words = text.split(Regex("\\s")).filter(String::isNotEmpty)
        if (words.size <= 3 && "@" !in text && "." !in text && "/" !in text) {
            val allCapped = words.all { it.firstOrNull()?.isUpperCase() == true }
            if (allCapped && words.size == 1 && text.length >= 3) return true
        }
        return false
    }

    private fun processSection(section: String, lines: List<String>, document: ResumeDocument) {
        when (section) {
            "summary" -> document.summary = lines.filter(String::isNotEmpty).joinToString(" ")
            "experience" -> document.experience = parseExperienceEntries(lines).toMutableList()
            "projects" -> document.projects = parseProjectEntries(lines).toMutableList()
            "education" -> document.education = parseEducationEntries(lines).toMutableList()
            "skills" -> document.skillCategories = parseSkillsSection(lines).toMutableList()
            else -> {
                val meaningful = lines.filter { it.isNotBlank() }
                if (meaningful.isNotEmpty()) {
                    document.additionalSections += AdditionalResumeSection(
                        title = section.uppercase(),
                        lines = meaningful.toMutableList(),
                    )
                }
            }
        }
    }

    private fun parseExperienceEntries(lines: List<String>): List<ExperienceEntry> {
        val entries = mutableListOf<ExperienceEntry>()
        var current: ExperienceEntry? = null
        var expectTitle = false
        for (line in lines) {
            if (line.isEmpty()) continue
            val bullet = asBullet(line)
            if (bullet != null) {
                current?.bullets?.add(bullet)
                expectTitle = false
                continue
            }
            when {
                current == null -> {
                    val (name, date) = splitNameAndDate(line)
                    current = ExperienceEntry(company = name, dateRange = date)
                    expectTitle = true
                }
                expectTitle -> {
                    val (title, location) = splitTitleAndLocation(line)
                    current.title = title
                    current.location = location
                    expectTitle = false
                }
                shouldContinuePreviousBullet(line, current) -> appendToLastBullet(line, current)
                else -> {
                    entries += current
                    val (name, date) = splitNameAndDate(line)
                    current = ExperienceEntry(company = name, dateRange = date)
                    expectTitle = true
                }
            }
        }
        current?.let(entries::add)
        return entries
    }

    private fun parseProjectEntries(lines: List<String>): List<ProjectEntry> {
        val entries = mutableListOf<ProjectEntry>()
        var current: ProjectEntry? = null
        for (line in lines) {
            if (line.isEmpty()) continue
            val bullet = asBullet(line)
            if (bullet != null) {
                if (current == null) current = ProjectEntry()
                current.bullets += bullet
                continue
            }
            if (shouldContinuePreviousProjectBullet(line, current)) {
                appendToLastProjectBullet(line, current!!)
                continue
            }
            current?.let(entries::add)
            val (name) = splitNameAndDate(line)
            current = ProjectEntry(name = name.ifEmpty { line })
        }
        current?.let(entries::add)
        return entries
    }

    private fun parseEducationEntries(lines: List<String>): List<EducationEntry> {
        val entries = mutableListOf<EducationEntry>()
        var current: EducationEntry? = null
        var expectDegree = false
        for (line in lines) {
            if (line.isEmpty()) continue
            val bullet = asBullet(line)
            if (bullet != null) {
                current?.details?.add(bullet)
                expectDegree = false
                continue
            }
            when {
                current == null -> {
                    val (institution, date) = splitNameAndDate(line)
                    current = EducationEntry(institution = institution, dateRange = date)
                    expectDegree = true
                }
                expectDegree -> {
                    val (degree, location) = splitTitleAndLocation(line)
                    current.degree = degree
                    current.location = location
                    expectDegree = false
                }
                else -> {
                    entries += current
                    val (institution, date) = splitNameAndDate(line)
                    current = EducationEntry(institution = institution, dateRange = date)
                    expectDegree = true
                }
            }
        }
        current?.let(entries::add)
        return entries
    }

    private fun parseSkillsSection(lines: List<String>): List<SkillCategory> {
        val categories = mutableListOf<SkillCategory>()
        for (line in lines) {
            if (line.isEmpty()) continue
            val colonIndex = line.indexOf(':')
            if (colonIndex >= 0) {
                val name = line.substring(0, colonIndex).trimHorizontalWhitespace()
                val items = line.substring(colonIndex + 1).trimHorizontalWhitespace()
                if (name.isNotEmpty() && items.isNotEmpty() && name.length < 40) {
                    categories += SkillCategory(name, items)
                    continue
                }
            }
            categories += SkillCategory(items = line)
        }
        return categories
    }

    private fun asBullet(line: String): String? {
        val trimmed = line.trimHorizontalWhitespace()
        val bulletChars = setOf('•', '·', '▪', '▸', '◦', '‣', '–', '—')
        if (trimmed.firstOrNull() in bulletChars) return trimmed.drop(1).trimHorizontalWhitespace()
        if (trimmed.startsWith("- ") || trimmed.startsWith("* ")) {
            return trimmed.drop(2).trimHorizontalWhitespace()
        }
        return null
    }

    private fun shouldContinuePreviousBullet(line: String, current: ExperienceEntry): Boolean =
        current.bullets.isNotEmpty() && splitNameAndDate(line).second.isEmpty()

    private fun appendToLastBullet(line: String, entry: ExperienceEntry) {
        if (entry.bullets.isEmpty()) return
        val continuation = line.trimHorizontalWhitespace()
        if (continuation.isNotEmpty()) {
            entry.bullets[entry.bullets.lastIndex] += " $continuation"
        }
    }

    private fun shouldContinuePreviousProjectBullet(line: String, current: ProjectEntry?): Boolean {
        if (current == null || current.bullets.isEmpty()) return false
        val trimmed = line.trimHorizontalWhitespace()
        if (trimmed.isEmpty() || splitNameAndDate(trimmed).second.isNotEmpty()) return false
        if (trimmed.length > 38) return true
        val first = trimmed.first()
        return first.isLowerCase() || first in "(),;:/"
    }

    private fun appendToLastProjectBullet(line: String, entry: ProjectEntry) {
        if (entry.bullets.isEmpty()) return
        val continuation = line.trimHorizontalWhitespace()
        if (continuation.isNotEmpty()) {
            entry.bullets[entry.bullets.lastIndex] += " $continuation"
        }
    }

    private fun splitNameAndDate(line: String): Pair<String, String> {
        val match = dateRegex.findAll(line).lastOrNull() ?: return line to ""
        val date = match.value.trimHorizontalWhitespace()
        val name = line.substring(0, match.range.first).trimHorizontalWhitespace()
        return if (name.isEmpty()) line to "" else name to date
    }

    private fun splitTitleAndLocation(line: String): Pair<String, String> {
        val words = line.split(Regex("\\s")).filter(String::isNotEmpty)
        if (words.size <= 1) return line to ""
        val last = words.last()
        if (last.lowercase() in knownLocations) {
            val title = words.dropLast(1).joinToString(" ")
            return (title.ifEmpty { line }) to last
        }
        if (words.size >= 3) {
            val lastTwo = words.takeLast(2).joinToString(" ")
            if (lastTwo.lowercase() in knownLocations) {
                val title = words.dropLast(2).joinToString(" ")
                return (title.ifEmpty { line }) to lastTwo
            }
        }
        val stateMatch = Regex(",\\s*[A-Z]{2}${'$'}").find(line)
        if (stateMatch != null) {
            val location = stateMatch.value.trim(',', ' ')
            val title = line.substring(0, stateMatch.range.first).trimHorizontalWhitespace()
            if (title.isNotEmpty()) return title to location
        }
        return line to ""
    }

    private fun splitPreservingEmpty(value: String): List<String> {
        val result = mutableListOf<String>()
        var start = 0
        for (index in value.indices) {
            if (value[index] == '\n') {
                result += value.substring(start, index)
                start = index + 1
            }
        }
        result += value.substring(start)
        return result
    }

    private fun String.trimHorizontalWhitespace(): String = trim { it.isHorizontalWhitespace() }

    private fun Char.isHorizontalWhitespace(): Boolean =
        isWhitespace() && this !in "\n\r\u000B\u000C\u0085\u2028\u2029"
}
